enclave-runner/src/usercalls/mod.rs

```diff
@@ -545,6 +565,37 @@ fn trap_attached_debugger(tcs: usize) {
     }
 }
 
+/// UsercallExtension provides a mechanism for the enclave code interface with an external service via a modified runner.
+/// If the implemenentation of the functions of UsercallExtension reutrn None, then [default functions](../../fortanix_sgx_abi/struct.Usercalls.html#networking)
+/// would be called which inteface the enclave with TCP stream.
+/// An implementation of UsercallExtension, can be registered while [building](../struct.EnclaveBuilder.html#method.usercall_ext) the enclave.
+
+#[allow(unused)]
+pub trait UsercallExtension : 'static + Send + Sync + std::fmt::Debug {
+    /// Allows the enclave to connect to an external service specified by the addr bytes.
+    /// The Runner should determine the service that the enclave is trying to connect to by parsing the addr bytes.
+    /// If `connect_stream` returns None, the default implementation of [connect_stream](../../fortanix_sgx_abi/struct.Usercalls.html#method.connect_stream) would be called.
+    /// Apart from connecting to the external service, the behavior of this function should be same as the default implementation.
```

Comment:
In practice, what should a service push to `local_addr` and `peer_addr`? Why are you pushing `"None"` in the example? (This might be answered in other docs, but I think it would be useful to put here).

Reply:
It makes more sense for [connect for a socket](https://edp.fortanix.com/docs/api/fortanix_sgx_abi/struct.Usercalls.html#method.connect_stream).
I am not sure what would be exact use case for generic API extension.
I will add the below part :
```
The enclave may optionally request the local or peer socket addresses be returned in local_addr or peer_addr, respectively. On success, if local_addr and/or peer_addr is not NULL, userspace will allocate memory for the address and populate the respective ByteBuffer appropriately.
```
Please suggest ways to make it more informative.

Reply:
Looking at Jethro's implementation and the Rust source code, the data we push to `local_addr` and `peer_addr` should be able to be parsed into a `SocketAddr`, see the following parsing code https://doc.rust-lang.org/src/std/net/parser.rs.html#262-297. So, I believe pushing the string `"None"` makes no sense.

If the custom usercall is not a connection to a socket (i.e. probably custom usercalls, since you can connect to a socket directly without a custom usercall), then you should ignore `local_addr` and `peer_addr` and not push anything, since the local address and peer address is irrelevant.

> Apart from connecting to the external service, the behavior of this function should be same as the default implementation.